File: windows/src/core/importParser.js
// Parses sipper://add-accounts links and native messaging payloads (docs/PROTOCOL.md).
// Ported from Sipper/Import/ImportParser.swift; the messages and rules match the Mac app.

import { makeAccount, newID, parseTransport, transportDefaultPort } from './models.js';

export const IMPORT_SCHEME = 'sipper';
export const IMPORT_ACTION = 'add-accounts';
export const MAX_PAYLOAD_BYTES = 512 * 1024;

export class ImportError extends Error {
  constructor(code, message, detail) {
    super(message);
    this.name = 'ImportError';
    this.code = code;
    this.detail = detail;
  }
}

const errors = {
  notAnImportURL: () => new ImportError('notAnImportURL', 'This link is not a Sipper import link.'),
  missingPayload: () => new ImportError('missingPayload', 'The import link has no payload.'),
  payloadTooLarge: () => new ImportError('payloadTooLarge', 'The import payload is too large.'),
  invalidBase64: () => new ImportError('invalidBase64', 'The import payload is not valid base64url.'),
  invalidJSON: (detail) => new ImportError('invalidJSON', `The import payload is not valid JSON (${detail}).`, detail),
  unsupportedVersion: (version) =>
    new ImportError('unsupportedVersion', `Import format version ${version} is not supported by this version of Sipper.`, version),
  noAccounts: () => new ImportError('noAccounts', 'The import contains no accounts.'),
};

/** Scheme, action and raw query of a URL string, without URLSearchParams' "+ means space" rule. */
function splitURL(link) {
  const match = /^([a-z][a-z0-9+.-]*):(.*)$/is.exec(String(link ?? '').trim());
  if (!match) return null;
  let rest = match[2];
  let query = '';
  const questionMark = rest.indexOf('?');
  if (questionMark >= 0) {
    query = rest.slice(questionMark + 1);
    rest = rest.slice(0, questionMark);
  }
  const hash = query.indexOf('#');
  if (hash >= 0) query = query.slice(0, hash);
  const action = rest.replace(/^\/+/, '').replace(/\/+$/, '').toLowerCase();
  return { scheme: match[1].toLowerCase(), action, query };
}

export function isImportURL(link) {
  const parts = splitURL(link);
  return Boolean(parts && parts.scheme === IMPORT_SCHEME && parts.action === IMPORT_ACTION);
}

function queryValue(query, name) {
  for (const pair of query.split('&')) {
    const equals = pair.indexOf('=');
    const key = equals >= 0 ? pair.slice(0, equals) : pair;
    if (key !== name) continue;
    const value = equals >= 0 ? pair.slice(equals + 1) : '';
    try {
      return decodeURIComponent(value);
    } catch {
      return value;
    }
  }
  return null;
}

export function decodeBase64URL(input) {
  let text = String(input).trim().replaceAll('-', '+').replaceAll('_', '/').replaceAll('=', '');
  if (!/^[A-Za-z0-9+/]*$/.test(text)) return null;
  const remainder = text.length % 4;
  if (remainder === 1) return null;
  if (remainder > 0) text += '='.repeat(4 - remainder);
  return Buffer.from(text, 'base64');
}

export function encodeBase64URL(bytes) {
  return Buffer.from(bytes).toString('base64').replaceAll('+', '-').replaceAll('/', '_').replaceAll('=', '');
}

/** Builds an import link for a JSON document (a string or bytes). */
export function makeImportURL(json) {
  return `${IMPORT_SCHEME}://${IMPORT_ACTION}?payload=${encodeBase64URL(Buffer.from(json))}`;
}

export function parseImportURL(link) {
  const parts = splitURL(link);
  if (!parts || parts.scheme !== IMPORT_SCHEME || parts.action !== IMPORT_ACTION) throw errors.notAnImportURL();
  const payload = queryValue(parts.query, 'payload');
  if (!payload) throw errors.missingPayload();
  if (Buffer.byteLength(payload) > MAX_PAYLOAD_BYTES * 2) throw errors.payloadTooLarge();
  const data = decodeBase64URL(payload);
  if (!data) throw errors.invalidBase64();
  return parseImportJSON(data);
}

export function parseImportJSON(data) {
  const bytes = typeof data === 'string' ? Buffer.from(data) : Buffer.from(data);
  if (bytes.length > MAX_PAYLOAD_BYTES) throw errors.payloadTooLarge();
  let document;
  try {
    document = JSON.parse(bytes.toString('utf8'));
  } catch (error) {
    throw errors.invalidJSON(error.message);
  }
  return importRequestFrom(document);
}

const quoted = (key) => `“${key}”`;

function expectOptionalObject(value, key) {
  if (value === undefined || value === null) return {};
  if (typeof value !== 'object' || Array.isArray(value)) throw errors.invalidJSON(`${quoted(key)} must be an object`);
  return value;
}

function expectOptionalString(object, key, path) {
  const value = object[key];
  if (value === undefined || value === null) return null;
  if (typeof value !== 'string') throw errors.invalidJSON(`${quoted(path ? `${path}.${key}` : key)} must be a string`);
  return value;
}

/** Validates the document's shape like the Mac app's Codable decoding does. */
export function importRequestFrom(document) {
  if (!document || typeof document !== 'object' || Array.isArray(document)) {
    throw errors.invalidJSON('the document must be an object');
  }
  if (!('version' in document)) throw errors.invalidJSON(`missing key ${quoted('version')}`);
  if (!Number.isInteger(document.version)) throw errors.invalidJSON(`${quoted('version')} must be a whole number`);
  if (!('accounts' in document)) throw errors.invalidJSON(`missing key ${quoted('accounts')}`);
  if (!Array.isArray(document.accounts)) throw errors.invalidJSON(`${quoted('accounts')} must be an array`);

  const source = expectOptionalObject(document.source, 'source');
  const profile = expectOptionalObject(document.profile, 'profile');
  const rawAccounts = document.accounts.map((account, index) => {
    if (!account || typeof account !== 'object' || Array.isArray(account)) {
      throw errors.invalidJSON(`account ${index + 1} must be an object`);
    }
    const path = `accounts[${index}]`;
    const fields = {};
    for (const key of ['label', 'displayName', 'username', 'authUsername', 'password', 'domain', 'server',
      'transport', 'callerIdName', 'callerIdNumber', 'voicemailNumber', 'notes']) {
      fields[key] = expectOptionalString(account, key, path);
    }
    if (account.port !== undefined && account.port !== null) {
      if (!Number.isInteger(account.port)) throw errors.invalidJSON(`${quoted(`${path}.port`)} must be a whole number`);
      fields.port = account.port;
    } else {
      fields.port = null;
    }
    return fields;
  });

  if (document.version !== 1) throw errors.unsupportedVersion(document.version);
  if (rawAccounts.length === 0) throw errors.noAccounts();

  const provider = clean(expectOptionalString(source, 'provider', 'source')) ?? 'manual';
  return {
    provider,
    sourceURL: clean(expectOptionalString(source, 'url', 'source')),
    sourceTitle: clean(expectOptionalString(source, 'title', 'source')),
    profileName: clean(expectOptionalString(profile, 'name', 'profile')),
    candidates: rawAccounts.map((raw) => candidateFrom(raw, provider)),
  };
}

function candidateFrom(raw, provider) {
  const validationErrors = [];
  const username = clean(raw.username) ?? '';
  const domain = clean(raw.domain) ?? '';
  // Passwords are kept verbatim; only an empty or whitespace-only password is invalid.
  const password = (raw.password ?? '').trim() === '' ? '' : raw.password;
  if (!username) validationErrors.push('Username is missing.');
  if (!domain) validationErrors.push('Domain is missing.');
  if (!password) validationErrors.push('Password is missing.');

  let transport = 'udp';
  const rawTransport = clean(raw.transport);
  if (rawTransport) {
    const parsed = parseTransport(rawTransport);
    if (parsed) transport = parsed;
    else validationErrors.push(`Unknown transport “${rawTransport}”.`);
  }

  let port = null;
  if (raw.port !== null) {
    if (raw.port >= 1 && raw.port <= 65535) {
      port = raw.port === transportDefaultPort(transport) ? null : raw.port;
    } else {
      validationErrors.push(`Port ${raw.port} is out of range.`);
    }
  }

  const server = clean(raw.server) ?? '';
  const authUsername = clean(raw.authUsername);
  const account = makeAccount({
    profileID: newID(),
    label: clean(raw.label) ?? '',
    displayName: clean(raw.displayName) ?? '',
    username,
    authUsername: authUsername && authUsername !== username ? authUsername : '',
    domain,
    server: server.toLowerCase() === domain.toLowerCase() ? '' : server,
    port,
    transport,
    voicemailNumber: clean(raw.voicemailNumber) ?? '*97',
    callerIDName: clean(raw.callerIdName) ?? '',
    callerIDNumber: clean(raw.callerIdNumber) ?? '',
    notes: clean(raw.notes) ?? '',
    source: provider,
  });
  return { id: newID(), account, password, validationErrors, existingAccountID: null };
}

export function clean(value) {
  if (typeof value !== 'string') return null;
  const trimmed = value.trim();
  return trimmed === '' ? null : trimmed;
}

File: Sipper/Import/ImportParser.swift
import Foundation

/// Wire format described in docs/PROTOCOL.md.
struct ImportDocument: Codable {
    struct Source: Codable {
        var provider: String?
        var url: String?
        var title: String?
    }

    struct ProfileHint: Codable {
        var name: String?
    }

    struct Account: Codable {
        var label: String?
        var displayName: String?
        var username: String?
        var authUsername: String?
        var password: String?
        var domain: String?
        var server: String?
        var port: Int?
        var transport: String?
        var callerIdName: String?
        var callerIdNumber: String?
        var voicemailNumber: String?
        var notes: String?
    }

    var version: Int
    var source: Source?
    var profile: ProfileHint?
    var accounts: [Account]
}

enum ImportError: LocalizedError, Equatable {
    case notAnImportURL
    case missingPayload
    case payloadTooLarge
    case invalidBase64
    case invalidJSON(String)
    case unsupportedVersion(Int)
    case noAccounts

    var errorDescription: String? {
        switch self {
        case .notAnImportURL: return "This link is not a Sipper import link."
        case .missingPayload: return "The import link has no payload."
        case .payloadTooLarge: return "The import payload is too large."
        case .invalidBase64: return "The import payload is not valid base64url."
        case .invalidJSON(let detail): return "The import payload is not valid JSON (\(detail))."
        case .unsupportedVersion(let v): return "Import format version \(v) is not supported by this version of Sipper."
        case .noAccounts: return "The import contains no accounts."
        }
    }
}

/// One account from an import, validated and ready to become a `SIPAccount`.
struct ImportCandidate: Identifiable, Hashable {
    let id = UUID()
    var account: SIPAccount
    var password: String
    var validationErrors: [String]
    /// Set by the app when an account with the same username and domain exists.
    var existingAccountID: UUID?

    var isValid: Bool { validationErrors.isEmpty }
    var isUpdate: Bool { existingAccountID != nil }
}

struct ImportRequest {
    var provider: String
    var sourceURL: String?
    var sourceTitle: String?
    var profileName: String?
    var candidates: [ImportCandidate]

    var validCount: Int { candidates.filter(\.isValid).count }
}

enum ImportParser {
    static let scheme = "sipper"
    static let action = "add-accounts"
    static let maxPayloadBytes = 512 * 1024

    static func isImportURL(_ url: URL) -> Bool {
        guard url.scheme?.lowercased() == scheme else { return false }
        return actionName(of: url) == action
    }

    private static func actionName(of url: URL) -> String {
        if let host = url.host, !host.isEmpty { return host.lowercased() }
        return url.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()
    }

    static func parse(url: URL) throws -> ImportRequest {
        guard isImportURL(url) else { throw ImportError.notAnImportURL }
        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        guard let payload = components?.queryItems?.first(where: { $0.name == "payload" })?.value,
              !payload.isEmpty else {
            throw ImportError.missingPayload
        }
        guard payload.utf8.count <= maxPayloadBytes * 2 else { throw ImportError.payloadTooLarge }
        guard let data = decodeBase64URL(payload) else { throw ImportError.invalidBase64 }
        return try parse(json: data)
    }

    static func parse(json data: Data) throws -> ImportRequest {
        guard data.count <= maxPayloadBytes else { throw ImportError.payloadTooLarge }
        let document: ImportDocument
        do {
            document = try JSONDecoder().decode(ImportDocument.self, from: data)
        } catch {
            throw ImportError.invalidJSON(describe(error))
        }
        return try request(from: document)
    }

    static func request(from document: ImportDocument) throws -> ImportRequest {
        guard document.version == 1 else { throw ImportError.unsupportedVersion(document.version) }
        guard !document.accounts.isEmpty else { throw ImportError.noAccounts }

        let provider = clean(document.source?.provider) ?? "manual"
        let candidates = document.accounts.map { candidate(from: $0, provider: provider) }
        return ImportRequest(provider: provider,
                             sourceURL: clean(document.source?.url),
                             sourceTitle: clean(document.source?.title),
                             profileName: clean(document.profile?.name),
                             candidates: candidates)
    }

    private static func candidate(from raw: ImportDocument.Account, provider: String) -> ImportCandidate {
        var errors: [String] = []

        let username = clean(raw.username) ?? ""
        let domain = clean(raw.domain) ?? ""
        // Passwords are kept verbatim; only an empty or whitespace-only password is invalid.
        let password = (raw.password ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "" : (raw.password ?? "")
        if username.isEmpty { errors.append("Username is missing.") }
        if domain.isEmpty { errors.append("Domain is missing.") }
        if password.isEmpty { errors.append("Password is missing.") }

        var transport = SIPTransport.udp
        if let rawTransport = clean(raw.transport) {
            if let parsed = SIPTransport(loosely: rawTransport) {
                transport = parsed
            } else {
                errors.append("Unknown transport “\(rawTransport)”.")
            }
        }

        var port: Int? = nil
        if let rawPort = raw.port {
            if (1...65535).contains(rawPort) {
                port = rawPort == transport.defaultPort ? nil : rawPort
            } else {
                errors.append("Port \(rawPort) is out of range.")
            }
        }

        let server = clean(raw.server) ?? ""
        let account = SIPAccount(
            profileID: UUID(),
            label: clean(raw.label) ?? "",
            displayName: clean(raw.displayName) ?? "",
            username: username,
            authUsername: clean(raw.authUsername).flatMap { $0 == username ? nil : $0 } ?? "",
            domain: domain,
            server: server.caseInsensitiveCompare(domain) == .orderedSame ? "" : server,
            port: port,
            transport: transport,
            voicemailNumber: clean(raw.voicemailNumber) ?? "*97",
            callerIDName: clean(raw.callerIdName) ?? "",
            callerIDNumber: clean(raw.callerIdNumber) ?? "",
            notes: clean(raw.notes) ?? "",
            source: provider
        )
        return ImportCandidate(account: account, password: password, validationErrors: errors)
    }

    // MARK: Helpers

    static func clean(_ value: String?) -> String? {
        guard let value else { return nil }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    static func decodeBase64URL(_ input: String) -> Data? {
        var s = input.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        s = s.replacingOccurrences(of: "=", with: "")
        let remainder = s.count % 4
        if remainder == 1 { return nil }
        if remainder > 0 { s += String(repeating: "=", count: 4 - remainder) }
        return Data(base64Encoded: s)
    }

    static func encodeBase64URL(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// Builds an import URL; used by tests and by the native messaging host.
    static func makeURL(jsonData: Data) -> URL? {
        var components = URLComponents()
        components.scheme = scheme
        components.host = action
        components.queryItems = [URLQueryItem(name: "payload", value: encodeBase64URL(jsonData))]
        return components.url
    }

    private static func describe(_ error: Error) -> String {
        if let decoding = error as? DecodingError {
            switch decoding {
            case .keyNotFound(let key, _): return "missing key “\(key.stringValue)”"
            case .typeMismatch(_, let ctx), .valueNotFound(_, let ctx), .dataCorrupted(let ctx):
                return ctx.debugDescription
            @unknown default: return String(describing: decoding)
            }
        }
        return error.localizedDescription
    }
}
